From d8e661ffac562d0db6b759861ee1472982ba220e Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Fri, 17 Jun 2016 16:37:06 -0700
Subject: [PATCH 686/827] drm: xilinx: dp: Set maximum swing / preemphsis level to 2 for ZynqMP

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

This patch is to combination of 7b8840fd and 16d76297.

The DP spec defines that the level 3 is optional to support for both
voltage swing and preemphasis. Thus, set the maximum voltage swing
and preemphasis level to 2 for ZynqMP DP subsystem.

Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/xilinx/xilinx_drm_dp.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
index a0c411d..3dfbf28 100644
--- a/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
+++ b/drivers/gpu/drm/xilinx/xilinx_drm_dp.c
@@ -454,6 +454,7 @@ static void xilinx_drm_dp_adjust_train(struct xilinx_drm_dp *dp,
 {
 	u8 *train_set = dp->train_set;
 	u8 voltage = 0, preemphasis = 0;
+	u8 max_voltage, max_preemphasis;
 	u8 i;
 
 	for (i = 0; i < dp->mode.lane_cnt; i++) {
@@ -467,11 +468,15 @@ static void xilinx_drm_dp_adjust_train(struct xilinx_drm_dp *dp,
 			preemphasis = p;
 	}
 
-	if (voltage >= DP_TRAIN_VOLTAGE_SWING_1200)
-		voltage |= DP_TRAIN_MAX_SWING_REACHED;
-
-	if (preemphasis >= DP_TRAIN_PRE_EMPHASIS_9_5)
-		preemphasis |= DP_TRAIN_MAX_PRE_EMPHASIS_REACHED;
+	max_voltage = (dp->dp_sub) ? DP_TRAIN_VOLTAGE_SWING_LEVEL_2 :
+				     DP_TRAIN_VOLTAGE_SWING_LEVEL_3;
+	if (voltage >= max_voltage)
+		voltage = max_voltage | DP_TRAIN_MAX_SWING_REACHED;
+	max_preemphasis = (dp->dp_sub) ? DP_TRAIN_PRE_EMPH_LEVEL_2 :
+					 DP_TRAIN_PRE_EMPH_LEVEL_3;
+	if (preemphasis >= max_preemphasis)
+		preemphasis = max_preemphasis |
+			      DP_TRAIN_MAX_PRE_EMPHASIS_REACHED;
 
 	for (i = 0; i < dp->mode.lane_cnt; i++)
 		train_set[i] = voltage | preemphasis;
-- 
2.9.3

